### PR TITLE
Add sanity plots

### PR DESCRIPTION
**Pull request by shashwatsridhar**
This PR adds sanity plots to the GUI to mimic the functionality of the Kilosort 2.0 GUI.

- A `SanityPlotWidget` class (in `gui/sanity_plots.py`) handles sanity plots created during spike sorting. 
- Two instances of the widget are created, for two sets of plots - dissimilarity matrices (end of clustering) and diagnostic plots (iteratively updated during learning). The corresponding functions are in `pykilosort/utils.py`.
- A custom color map used in these plots is also stored in `gui/palettes.py`. 
- The signatures of core functions in `learn.py` and `cluster.py` have also been changed to accommodate for the plots. This should not break anything else in the program, though.
- A checkbox is provided near the run buttons in the main window, so the user can decide whether or not to show sanity plots.

A screenshot of the final plots after successfully sorting a small dataset:

sanity_plot_example

### CODE REVIEW

File: pykilosort/gui/sanity_plots.py [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+
+        self.setWindowTitle(title)
+
+        self.create_remote_views()
+        self.arrange_views()
+
+        self.hide()
+
+    def create_remote_views(self):
+        for i in range(self.num_remote_plots):
```
Comment (alexmorley) [inline pykilosort/gui/sanity_plots.py]:
```suggestion
        for _ in range(self.num_remote_plots):
```

File: pykilosort/gui/sanity_plots.py [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+
+    def add_scatter(self,
+                    x_data: np.ndarray,
+                    y_data: np.ndarray,
+                    plot_pos: int,
+                    labels: dict,
+                    x_lim: t.Union[tuple, None] = None,
+                    y_lim: t.Union[tuple, None] = None,
+                    semi_log_x: t.Union[bool, None] = None,
+                    semi_log_y: t.Union[bool, None] = None,
```
Comment (alexmorley) [inline pykilosort/gui/sanity_plots.py]:
```suggestion
                    x_lim: t.Optional[tuple] = None,
                    y_lim: t.Optional[tuple] = None,
                    semi_log_x: t.Optional[bool] = None,
                    semi_log_y: t.Optional[bool] = None,
```

File: pykilosort/gui/sanity_plots.py [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+
+    def add_curve(self,
+                  x_data: np.ndarray,
+                  y_data: np.ndarray,
+                  plot_pos: int,
+                  labels: dict,
+                  x_lim: t.Union[tuple, None] = None,
+                  y_lim: t.Union[tuple, None] = None,
+                  semi_log_x: t.Union[bool, None] = None,
+                  semi_log_y: t.Union[bool, None] = None,
```
Comment (alexmorley) [inline pykilosort/gui/sanity_plots.py]:
```suggestion
                  x_lim: t.Optional[tuple] = None,
                  y_lim: t.Optional[tuple] = None,
                  semi_log_x: t.Optional[bool] = None,
                  semi_log_y: t.Optional[bool] = None,
```

File: pykilosort/learn.py [hunk 4]
```diff
@@ -1198,6 +1198,10 @@ def learnAndSolve8b(ctx):
                 *ndrop
             )
 
+            if sanity_plots:
+                assert plot_widgets is not None
```
Comment (alexmorley) [inline pykilosort/learn.py]:
```suggestion
                assert plot_widgets is not None, "if sanity_plots is set then plot_widgets cannot be None"
```

Comment (alexmorley) [reply pykilosort/learn.py]: ☝🏼 will give a better error message.

**Review by alexmorley — COMMENTED** [review]
few small suggestions

File: pykilosort/learn.py (hunk 4 above)
Comment (shashwatsridhar) [reply pykilosort/learn.py]: Nice! I didn't know of this nifty trick, thanks!

**Review by alexmorley — APPROVED** [review]

**Review by alexmorley — APPROVED** [review]

**Comment by shashwatsridhar** [comment]
Made the changes requested by @marius10p and incorporated the changes suggested by @alexmorley. The plots look like this now:

update_sanity_plots

The colormaps for all plots are better, imo. The spike amplitude vs. spike count plot (bottom right plot in the right window) now is a semilogx plot, like in the matlab version.

**Comment by marius10p** [comment]
Looks, good, but can you please make sure the vmin and vmax are the same abs value in the diagnostic plots? That way the background will be white, because it should be exactly 0.

**Comment by shashwatsridhar** [comment]
Thanks for the hint Marius, that steered me in the right direction with the colormap values. I wasn't aware of the vMin and vMax params from matlab (and subsequently matplotlib). They don't seem to have an obvious equivalent in pyqtgraph. Pyqtgraph follows a slightly different (and slightly confusing) terminology. I had manually set the params to get the "best looking" plots, but thanks to your hint, I saw how the data was already normalized, making it easier to choose the colormap parameters.

Here's how it looks now:

updated_sanity_plot_2

If this looks good, I'll push the changes and merge the PR.

**Pull request merged**